### ISSUE TITLE
gensim==0.12.4 mallet wrapper is not compatible with mallet-2.0.7

### ISSUE THREAD

**Issue by janrygl** [RaRe-Technologies/gensim#599]
Following error is raised for mallet-2.0.7. In mallet-2.0.8RC3, everything passes without any problems.

```python
for doc_no, (topics, stats) in enumerate(zip(model.load_document_topics(), docs.doc_stats)):
  File "/software/p27/local/lib/python2.7/site-packages/gensim/models/wrappers/ldamallet.py", line 271, in read_doctopics
    raise RuntimeError("invalid doc topics format at line %i in %s" % (lineno + 1, fname))
RuntimeError: invalid doc topics format at line 3 in /tmp/4303cb_doctopics.txt
```

```bash
$ head -n 5 /tmp/4303cb_doctopics.txt
#doc name topic proportion ...
0	0	0	0.9378669927812985	9	0.006903667468744601	8	0.006903667468744601	7	0.006903667468744601	6	0.006903667468744601	5	0.006903667468744601	40.006903667468744601	3	0.006903667468744601	2	0.006903667468744601	1	0.006903667468744601	
1	1	0	1.0780612802674239	9	0.005575655428533364	8	0.005575655428533364	7	0.005575655428533364	6	0.005575655428533364	5	0.005575655428533364	30.005575655428533364	2	0.005575655428533364	1	0.005575655428533364	
2	2	0	0.9184413079632608	9	0.009062076892971008	8	0.009062076892971008	7	0.009062076892971008	6	0.009062076892971008	5	0.009062076892971008	40.009062076892971008	3	0.009062076892971008	2	0.009062076892971008	1	0.009062076892971008	
3	3	0	0.9184413079632608	9	0.009062076892971008	8	0.009062076892971008	7	0.009062076892971008	6	0.009062076892971008	5	0.009062076892971008	40.009062076892971008	3	0.009062076892971008	2	0.009062076892971008	1	0.009062076892971008	
```

**Comment by piskvorky** [comment]
Looks like a bug introduced in e454e8c82767e776d97fbd1dafba39f14d048598 CC @jeiranj .

My guess is, in the old (2.0.7) sparse mallet format, not all topics are necessarily listed. Also seems to be true in @janrygl 's example `head` output.

The [original code for 2.0.7](
https://github.com/piskvorky/gensim/commit/e454e8c82767e776d97fbd1dafba39f14d048598#diff-220ca1b82d4f1c42d8f4432ff3a2a6ccL50) doesn't seem to have this constraint of "length == 2 * topics".

**Comment by cscorley** [comment]
That would be my fault for letting that through review for PR #516. I did test it against some sample 2.0.7 output to test against my comments, but I'm guessing that output wasn't sparse either. If 2.0.8 is always going to be dense, then the check for "length == num_topics" is all that is needed for that and to make sure there's an even number otherwise, e.g.,:

(changing [these lines](https://github.com/piskvorky/gensim/blob/5c2018964f5eda19970417a368689f650860e4cb/gensim/models/wrappers/ldamallet.py#L261-L271))
```python
                if len(parts) == self.num_topics:  # 2.0.8 dense format
                    doc = [(id_, weight)
                           for id_, weight in enumerate(map(float, parts))
                           if abs(weight) > eps]
                elif len(parts) % 2 == 0:  # 2.0.7 sparse format
                    doc = [(id_, weight)
                           for id_, weight in zip(map(int, parts[::2]),
                                                  map(float, parts[1::2]))
                           if abs(weight) > eps]
                else:
                    raise RuntimeError("invalid doc topics format at line %i in %s" % (lineno + 1, fname))
```

**Comment by piskvorky** [comment]
I'm a little worried there's some small room for confusion when the output was sparse (old mallet) and happens to be half the number of topics (so that `len(parts) == self.num_topics` is triggered, but shouldn't be).

If there's no bulletproof way of guessing the Mallet version, then I'd prefer to fail explicitly, rather than guess (potentially wrong).

**Comment by cscorley** [comment]
You're right, that's an issue.  Another option is to check the couple lines for a few (int, float) parse-able pairs, and decide from there on success/fail.

**Comment by piskvorky** [comment]
Ah yes, I think that could work reliably. A bulletproof guess would be preferable to bugging the user with extra params, of course :)

**Comment by tmylk** [comment]
Or we could just read the Mallet version directly from pom.xml file in mallet path.
It is stored as `<version>2.0.8-SNAPSHOT</version>`. What do you think?

I will be able to implement and release it after 15 Feb.

**Comment by RishabGoel** [comment]
@janrygl : I am trying to handle the above error. Could you please refer the parameters you used, while running gensim ??

**Comment by janrygl** [comment]
@RishabGoel I am sorry but I don't remember exactly, I can guess some parameters from the source code, see following:

LdaMallet initialization:
```python
num_topics=100
workers=7
prefix='/tmp/34f883_'
optimize_interval=10
iterations=400..1000  # depends on doc count
id2word=gensim.corpora.Dictionary
```

Training:
```bash
/path/to/mallet train-topics --input self.fcorpusmallet() --num-topics 100 
  --optimize-interval 10
  --num-threads 7 --output-state self.fstate() --output-doc-topics self.fdoctopics()
  --output-topic-keys self.ftopickeys()
  --num-iterations 400..1000 --inferencer-filename  self.finferencer() 
  --evaluator-filename e.eval --optimize-burn-in 100
  --random-seed 1
```

**Comment by tmylk** [comment]
@cscorley @RishabGoel has implemented good ways to find out Mallet version so trying to re-cap what is the best solution here.

Mallet 2.0.8 is always dense. 

Mallet 2.0.7 can be either dense or sparse. 

What is the flag that makes 2.0.7 sparse? @RishabGoel has been unable to reproduce the above output.

**Comment by cscorley** [comment]
@tmylk @RishabGoel I've never gotten sparse output from Mallet.  However, I'd think setting `--doc-topics-threshold` would do it.

**Comment by RishabGoel** [comment]
I also feel the same thing. But by default the threshold is set to 0.0. I have checked the source code of mallet2.0.7.And as stated above by @janrygl  there is not such threshold set while running the mallet. So, might be, in the   mallet that @janrygl  used such threshold was set to some value.

**Comment by tmylk** [comment]
Fixed in #664
